### ISSUE TITLE
HTTP2 support

### ISSUE THREAD

**Issue by hisco** [olalonde/follow-redirects#93]
With http2 already being a very standard protocol, it would be nice to support http2 redirects.

It was very easy to combine [http2-client](https://www.npmjs.com/package/http2-client) and `follow-redirects` to accomplish exactly that:
[h2-follow-redirects](https://github.com/hisco/h2-follow-redirects
)
```js
const http2Client = require('http2-client');
const followRedirects = require('follow-redirects');
const h2Follow = require('follow-redirects').wrap({
  https : http2Client
});

module.exports = {
  http2 : h2Follow.https,
  http  : followRedirects.http,
  https : followRedirects.https
};

```

Please let me know if you would like me to generate a pull request with a similar code.

I would expect it to be turned on manually something like:
```js
require('follow-redirects').enableHttp2(); //<-- http2 dependencies will be loaded
require('follow-redirects').http2 //<-- http2 is now available
```

**Comment by RubenVerborgh** [comment]
> it would be nice to support http2 redirects.

Definitely!

> I would expect it to be turned on manually

Why not just by default?

**Comment by EDtoro** [comment]
While http2 is a standard protocol it's still considered experimental by the Node.js team.
With that being said it's have been like that for a quite a while and the API is stable - I feel very comfortable working with it.

I'm just not sure everyone will feel the same way. Therefore, having the option to manually load it would be a nice solution to any usage.

**Comment by hisco** [comment]
While http2 is a standard protocol it's still considered experimental by the Node.js team.
With that being said it's have been like that for a quite a while and the API is stable - I feel very comfortable working with it.

I'm just not sure everyone will feel the same way. Therefore, having the option to manually load it would be a nice solution to any usage.

**Comment by hisco** [comment]
BTW I can also use a getter that will load it only once and cache it.
As a result the user wouldn't notice that the module will be loaded only when used...

**Comment by RubenVerborgh** [comment]
> While http2 is a standard protocol it's still considered experimental by the Node.js team.

That doesn't really matter; we can just expose it.

The important thing here is mostly how we decide which one to use. If we encounter an `https:` URL (in the beginning or during a redirect), how do we decide whether to use HTTP or HTTP/2? I don't know the http2 API, is there functionality for this, or does it automatically downgrade?

**Comment by RubenVerborgh** [comment]
Actually, thinking about this some more, HTTP/2 is a different beast altogether.

I see few benefits from using the [compatibility API](https://nodejs.org/api/http2.html#http2_compatibility_api), which largely treats HTTP/2 as ordinary HTTP (minus some compression). If we really want HTTP/2, we should adopt the new API, but that is a different thing altogether.

What benefits do you see of using HTTP/2 in this way?

**Comment by RubenVerborgh** [comment]
Ah, http2-client is the piece of software that falls back from HTTP/2 to HTTP, that was the important piece of knowledge missing here :smile: So this interface compatibility we need is provided by http2-client.

> Benefits are similar to browsers + support:
> 
>     Binary, multiplexing, etc...

Well, not necessarily. If I use `http2-client` through `follow-redirects`, will I have multiplexing? I.e., does `http2-client` do automatic multiplexing across different requests? Because, given the HTTP-compatible API it uses, that might not be the case.

If the redirect is to the same server, will the connection be reused automatically? Because that would be a necessary condition for the HTTP/2 optimizations to happen.
(If the redirect is to another server, then we do not really benefit anyway.)

We have binary in any case with HTTPS.

>     Some servers may provide better or only http2 support in the future.

That would indeed be a reason, but I don't see any evidence for this at the moment.

**Comment by hisco** [comment]
> http2-client is the piece of software that falls back from HTTP/2 to HTTP

That is correct.

> does http2-client do automatic multiplexing across different requests

Yes

>  Because, given the HTTP-compatible API it uses, that might not be the case

It's not using the compatibility API.

> If the redirect is to the same server, will the connection be reused automatically? Because that would be a necessary condition for the HTTP/2 optimizations to happen.
(If the redirect is to another server, then we do not really benefit anyway.)

That is correct, if it's the only request you are doing concurrently. However, some software may have few requests at once:
```js
  http2Client.get('http2-endpoint');
  http2Client.get('http2-endpoint');
  http2Client.get('http2-endpoint');
  http2Client.get('http2-endpoint');
```
Only a single http2 connection will be opened.

> That would indeed be a reason, but I don't see any evidence for this at the moment.

I agree on the frontend. However, for internal usage some backend only have internals clients and would like to implement only http2 support because it's an improved transportation protocol.
You can consider gRPC case which only supports http2.

As a general note I would like to see all major node modules that handles http transportation to support http2 in addition (Both client & server).

**Comment by hisco** [comment]
> so we need to be very picky about any dependencies we pull

True, at the moment I can only suggest; use it in a (npm)beta version and add you as collaborator.

> Which projects is it used in a the moment

I'm aware only of some private projects I have.

> prove itself a bit more first

Sure, ping this thread when it's time 😄
